File: 0109-convert-sorted-list-to-binary-search-tree/0109-convert-sorted-list-to-binary-search-tree.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {ListNode} head
 * @return {TreeNode}
 */
var sortedListToBST = function(head) {
  if (!head) return null;
  if (!head.next) return new TreeNode(head.val);
  
  let [first, second, fast] = [null, head, head];
  while (fast && fast.next) {
    first = second;
    second = second.next;
    fast = fast.next.next;
  }
  
  first.next = null;
  const node = new TreeNode(second.val);
  node.left = sortedListToBST(head);
  node.right = sortedListToBST(second.next);
  
  return node;
};

File: 0109-convert-sorted-list-to-binary-search-tree/0109-convert-sorted-list-to-binary-search-tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
        if not head:
            return None
        if not head.next:
            return TreeNode(head.val)
        first, second, fast = None, head, head
        while fast and fast.next:
            first = second
            second = second.next
            fast = fast.next.next
        first.next = None
        
        node = TreeNode(second.val)
        node.left = self.sortedListToBST(head)
        node.right = self.sortedListToBST(second.next)
        return node
            
        